vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt

```diff
@@ -101,14 +122,104 @@ class VesselImpl(
+     *
+     * @return true only if the entire database could be loaded into cache in the allotted timeout
+     */
+    private fun preloadImpl(timeoutMS: Int?): Boolean {
+        val cursor = db.query("SELECT * FROM vessel", emptyArray())
+
+        cursor.use {
+            val typeCol = cursor.getColumnIndex("type")
+            val dataCol = cursor.getColumnIndex("data")
+
+            val startTimeMS = System.currentTimeMillis()
+            while (cursor.moveToNext()) {
+                val type = cursor.getString(typeCol)
+                val data = cursor.getString(dataCol)
+
+                val kclass = try {
+                    Class.forName(type).kotlin
+                } catch (error: Exception) {
+                    continue
+                }
+
+                val obj = if (data != null) fromJson(data, kclass) else nullValue
+                cache?.set(type, obj as Any)
+
+                timeoutMS?.let {
+                    val duration = System.currentTimeMillis() - startTimeMS
+                    if (duration > it) {
+                        cursor.close()
+                        return false
+                    }
+                }
+            }
+
+            return true
+        }
+    }
+
+    override suspend fun preload(timeoutMS: Int?) {
+        if (cache == null) {
+            return
+        }
+
+        val loadCompleted = profiler.time(Span.PRELOAD_FROM_DB) {
+            /**
+             * This is exactly the same pattern that the Room code generator emits/
+             * Implementing this way so preloading can bail out if a time limit is hit
+             *
+             * Note:  This may run on a different dispatcher if running in a transaction, meaning
+             * the profiling may be a little off with respect to which coroutine is doing the work.
+             * Deeming this as acceptable - the profiled execution time is still valid
+             */
+            CoroutinesRoom.execute(db, false) {
+                preloadImpl(timeoutMS)
+            }
+        }
+
+        if (!loadCompleted) {
+            profiler.count(Event.PRELOAD_TIMEOUT)
+        }
+    }
+
+    override fun preloadBlocking(timeoutMS: Int?) {
+        if (cache == null) {
+            return
+        }
+
+        val loadCompleted = profiler.timeBlocking(Span.PRELOAD_FROM_DB) {
+            preloadImpl(timeoutMS)
+        }
+
+        if (!loadCompleted) {
+            profiler.countBlocking(Event.PRELOAD_TIMEOUT)
+        }
+    }
+
+    // endregion
+
+    // region profiling
+
+    override val profileData
+        get() = profiler.snapshot
+
     // endregion
 
     // region helper functions
 
     /**
      * Convert a stored json string into the specified data type.
      */
-    private fun <T : Any> fromJson(value: String, type: KClass<T>): T? = gson.fromJson<T>(value, type.java)
+    private fun <T : Any> fromJson(value: String, type: KClass<T>): T? = gson.fromJson(value, type.java)
```

Comment:
IIRC the IDE said this generic parameter wasn't required, but we were getting exceptions without it

Reply:
At runtime?  How?  Generics are type erased.